### PR TITLE
Further optimize Eval.fileModifiedHash with nio

### PR DESCRIPTION
**Pull request by dwijnand**
working off of @fommil's info in https://github.com/sbt/sbt/pull/4067#issuecomment-380580693.

haven't benchmarked this

### CODE REVIEW

File: main-actions/src/main/scala/sbt/compiler/Eval.scala [hunk 1]
```diff
@@ -503,25 +505,26 @@ private[sbt] object Eval {
...
-    if (f.isDirectory)
-      (f listFiles classDirFilter) foreach { x =>
-        fileModifiedHash(x, digester)
-      } else digester.update(bytes(getModifiedTimeOrZero(f)))
-
-    digester.update(bytes(f.getAbsolutePath))
+    Files.walkFileTree(
+      f.toPath,
+      new SimpleFileVisitor[Path] {
+        override def visitFile(file: Path, attrs: BasicFileAttributes) = {
```
Comment (fommil) [inline main-actions/src/main/scala/sbt/compiler/Eval.scala]: from memory you can use `visitDirectory` to get the `attrs` of all the files in the dir.

File: main-actions/src/main/scala/sbt/compiler/Eval.scala [hunk 2]
```diff
@@ -503,25 +505,26 @@ private[sbt] object Eval {
...
+          if (classDirFilter accept file.toFile)
+            digester.update(bytes(attrs.lastModifiedTime.toMillis))
+          digester.update(bytes(file.toFile.getAbsolutePath))
```
Comment (fommil) [inline main-actions/src/main/scala/sbt/compiler/Eval.scala]: there should be a way to get the absolute path without going via File... see https://github.com/ensime/ensime-server/blob/3.0/api/src/main/scala/org/ensime/io/Canon.scala#L31-L42

Comment (fommil) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: also, avoiding two `.toFile` calls would be good here

Comment (dwijnand) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: is toFile expensive?

File: main-actions/src/main/scala/sbt/compiler/Eval.scala (hunk 1 above)
Comment (dwijnand) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: I don't see how to do that. preVisitDirectory only has one attrs, with one lastModifiedTime for the directory itself.

File: main-actions/src/main/scala/sbt/compiler/Eval.scala (hunk 2 above)
Comment (fommil) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: `Path` uses an optimised hierarchical representation, whereas `File` constructs a full `String`

when I've profiled sbt on large projects, this has been the single biggest cause of heap usage.

Comment (dwijnand) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: the alternative I found is `dir.toAbsolutePath.toString`

Comment (fommil) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: it's a shame that you need to construct the string for the callback... I guess **that** optimisation needs to wait until the API can be changed.

Comment (dwijnand) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: re-implemented!

Comment (dwijnand) [reply main-actions/src/main/scala/sbt/compiler/Eval.scala]: oh you mean the other string

**Comment by dwijnand** [comment]
moving to ticket https://github.com/sbt/sbt/issues/4116.

**Pull request closed without merging**